src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1120,15 +1132,35 @@ private void OnFileChanged(object sender, FileSystemEventArgs e)
             // - the host.json file was changed
             // - a function.json file was changed
             // - a function directory was added/removed/renamed
+            // A full host shutdown is performed when an assembly (.dll, .exe) in a watched directory is modified
             string fileName = Path.GetFileName(e.Name);
             if (isWatchedDirectory ||
                 ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 !_directorySnapshot.SequenceEqual(Directory.EnumerateDirectories(ScriptConfig.RootScriptPath)))
             {
-                // a host level configuration change has been made which requires a
-                // host restart
-                _restart(e);
+                bool shutdown = false;
+                string fileExtension = Path.GetExtension(fileName);
+                if (string.Equals(fileExtension, ".dll", StringComparison.OrdinalIgnoreCase) || string.Equals(fileExtension, ".exe", StringComparison.OrdinalIgnoreCase))
+                {
+                    shutdown = true;
+                }
+
+                TraceFileChangeRestart(e.ChangeType.ToString(), e.FullPath, shutdown);
+                ScheduleRestart(shutdown);
+            }
+        }
+
+        private void ScheduleRestart(bool shutdown)
```

Comment:
This might be a bit confusing, but `_shutdown` is debounced (as is `_restart`) and calling the delegate is essentially scheduling that execution to after the debounce window. This method is named to reflect that and it sets the flag indicating that a shutdown was requested so it can be checked as calls to restart could be made within the debounce window.